lib/rucio/common/types.py

```diff
@@ -99,3 +102,58 @@ class InternalScope(InternalType):
     '''
     def __init__(self, scope, vo='def', fromExternal=True):
         super(InternalScope, self).__init__(value=scope, vo=vo, fromExternal=fromExternal)
+
+
+class RSEDomainLANDict(TypedDict):
+    read: Optional[int]
+    write: Optional[int]
+    delete: Optional[int]
+
+
+class RSEDomainWANDict(TypedDict):
+    read: Optional[int]
+    write: Optional[int]
+    delete: Optional[int]
+    third_party_copy_read: Optional[int]
+    third_party_copy_write: Optional[int]
+
+
+class RSEDomainsDict(TypedDict):
+    lan: RSEDomainLANDict
+    wan: RSEDomainWANDict
+
+
+class RSEProtocolDict(TypedDict):
+    auth_token: Optional[str]
+    hostname: str
+    scheme: str
+    port: int
+    prefix: str
+    impl: str
+    domains: RSEDomainsDict
+    extended_attributes: Optional[Union[str, dict[str, Any]]]
+
+
+class RSESettingsDict(TypedDict):
+    availability_delete: Optional[bool]
+    availability_read: Optional[bool]
+    availability_write: Optional[bool]
+    credentials: Optional[str]
+    lfn2pfn_algorithm: str
+    qos_class: Optional[str]
+    staging_area: bool
+    rse_type: Optional[str]
+    sign_url: Optional[str]
+    read_protocol: int
+    write_protocol: int
+    delete_protocol: int
+    third_party_copy_read_protocol: int
+    third_party_copy_write_protocol: int
+    id: str
+    rse: str
+    type: str
+    volatile: bool
+    verify_checksum: bool
+    deterministic: bool
+    domain: str
+    protocols: list[RSEProtocolDict]
```

Comment:
Consider this real-life example:

```python
>>> pprint(get_rse_protocols('a8cbe2f23cc44cac8417df73d259d392'))
{'availability_delete': True,
 'availability_read': True,
 'availability_write': True,
 'credentials': None,
 'delete_protocol': 1,
 'deterministic': False,
 'domain': ['lan', 'wan'],
 'id': 'a8cbe2f23cc44cac8417df73d259d392',
 'lfn2pfn_algorithm': 'hash',
 'protocols': [{'domains': {'lan': {'delete': 1, 'read': 1, 'write': 1},
                            'wan': {'delete': 1,
                                    'read': 1,
                                    'third_party_copy_read': 1,
                                    'third_party_copy_write': 1,
                                    'write': 1}},
                'extended_attributes': {'space_token': 'ATLASDATATAPE',
                                        'web_service_path': '/srm/managerv2?SFN='},
                'hostname': 'ccsrm.in2p3.fr',
                'impl': 'rucio.rse.protocols.gfal.Default',
                'port': 8443,
                'prefix': '/pnfs/in2p3.fr/data/atlas/atlasdatatape/',
                'scheme': 'srm'}],
 'qos_class': None,
 'read_protocol': 1,
 'rse': 'IN2P3-CC_DATATAPE',
 'rse_type': 'TAPE',
 'sign_url': None,
 'staging_area': False,
 'third_party_copy_read_protocol': 1,
 'third_party_copy_write_protocol': 1,
 'verify_checksum': True,
 'volatile': False,
 'write_protocol': 1}
 ```

I can see some deviations from the hints you’ve provided:

 * `availability_*` may not be `None`.
 * `domain` is a `list[str]`.
 * `type` does not exist. It’s there with `get_rse` but it’s an `RSEType`, not an `str`. However, that structure looks *very* different.
```python
>>> pprint(get_rse('a8cbe2f23cc44cac8417df73d259d392'))
{'ASN': None,
 'ISP': None,
 'availability': 7,
 'availability_delete': True,
 'availability_read': True,
 'availability_write': True,
 'city': None,
 'continent': None,
 'country_name': None,
 'created_at': datetime.datetime(2014, 3, 17, 10, 22, 27),
 'deleted': False,
 'deleted_at': None,
 'deterministic': False,
 'id': 'a8cbe2f23cc44cac8417df73d259d392',
 'latitude': 45.7825,
 'longitude': 4.8655,
 'qos_class': None,
 'region_code': None,
 'rse': 'IN2P3-CC_DATATAPE',
 'rse_type': <RSEType.TAPE: 'TAPE'>,
 'staging_area': False,
 'time_zone': None,
 'type': <RSEType.TAPE: 'TAPE'>,
 'updated_at': datetime.datetime(2023, 6, 27, 16, 0, 7),
 'vo': 'def',
 'volatile': False}
```

We might need two another class to do this properly for now, but I wonder whether they can be merged in the future.

Reply:
> So we need to create class for get_rse and RSEType class?

No, I would suggest sticking to `get_rse_protocols()` for this PR. You’re doing good work!

Reply:
Forgive me, I might have confused you. For this PR, `rse_type` is a `str`, as seen in https://github.com/rucio/rucio/blob/7f0d229ac0b3bc7dec12c6e158bea2b82d414a3b/lib/rucio/core/rse.py#L1410.

Reply:
Ok, then I'll fix it now

> Forgive me, I might have confused you. For this PR, `rse_type` is a `str`, as seen in
> 
> https://github.com/rucio/rucio/blob/7f0d229ac0b3bc7dec12c6e158bea2b82d414a3b/lib/rucio/core/rse.py#L1410
> .

Reply:
Any other fixes need to be done or this PR should stay as a draft for now?